src/x_system/x_setup.1
.TH X_SETUP 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
\fBx_setup\fP \- Determine Pairs of Legs that need Cross-Over Checking
.SH SYNOPSIS
\fBx_setup\fP [ \fB\-L\fP\fIlegfile\fP ] [ \fB\-R\fP\fIwest/east/south/north\fP ]
.SH DESCRIPTION
\fBx_setup\fP scans the \fIgmtindex.b\fP file that contains information on which legs occupy each 1 by
1 degree bin and reports all the pairs of legs that may overlap. Each pair is only reported once, even if
the two legs involved are found in other bins also. 
.TP
\fB\-L\fP
Only report pairs where at least one of the legs are present in the \fIlegfile\fP.
.TP
\fB\-R\fP
The region of interest.  [Default is world]
.SH BEWARE
The \fB\-L\fP option is useful when one wants to check out a handful of new cruises. Obviously we
only need to compute COEs for the pairs that involve one or two of the new cruises, since all the
other combinations have already been done.
.SH SEE ALSO
.IR GMT (1),
.IR x_system (1)
.SH REFERENCES
Wessel, P. XOVER: A Cross-over Error Detector for Track Data,
\fIComputers & Geosciences, 15\fP, 333-346.


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/x_system/x_setup.c
/*	$Id: x_setup.c 9545 2011-07-27 19:31:54Z pwessel $
 *
 * XSETUP will read the gmtindex files and create a list of
 * pairs of legs that cross the same bin. As an option, the
 * user may specify a file with a list of legs. Then, only
 * those legs and whatever legs they cross are listed. This
 * is the usual way to do things when a new batch of legs
 * have been added to the databank. !! These new legs must
 * first be run through binlegs so that xsetup can find their
 * binindices.
 *
 * Author:	Paul Wessel
 * Date:	15-NOV-1987
 * Revised:	18-FEB-1989
 * 		15-FEB-1990	PW: Fixed bug when -L is used
 * 		27-MAR-1992	PW: Updated to new indexfile format
 * 		06-MAR-2000	PW: POSIX
 *				gmt_legs.d and gmt_index.b are now in GMT_SHAREDIR/mgg
 *		17-JUL-2000	Replace MAX by MAXLEGS and use gmt.h
 *
 */
#include "gmt.h"
#include "x_system.h"

#define MAXLEGS 64000	/* Max 32000 inside and 32000 outside legs */
#define MAXNEW 500	/* Max # new legs to check [-L option] */
#define NTOT 10000	/* Max # legs in database */

struct INFO {
	char legname[16], use;
	GMT_LONG leg_id, seq_no;
	unsigned char *leglist;
};

struct LEGNAMES {
	GMT_LONG leg_no;
	char gmt;
} leg_info[NTOT];

struct INFO *make_info (char *name, GMT_LONG id_no);
GMT_LONG findleg (char *name);

GMT_LONG nlegs_used = 0;
char leg_used[MAXNEW][10];
GMT_LONG legpointer[NTOT];
struct INFO *ptr[MAXLEGS];

int main (int argc, char *argv[])
{
	GMT_LONG i, j, byte_1, byte_2, bit_1, bit_2, leg_1, leg_2, ok, no, n_tot_legs;
	GMT_LONG lat, lon, west = 0, east = 360, south = -90, north = 89;
	long n_alloc;
	int bin, nlegs, record;	/* Let these remain 32-bit */
	double w, e, s, n;
	char lname[8], line[BUFSIZ];
	unsigned char turn_on[8];
	FILE *fleg = NULL, *fbin = NULL, *fpl = NULL;
	GMT_LONG error = FALSE, skip;
	size_t not_used = 0;

	argc = (int) GMT_begin (argc, argv);

	for (i = 1; i < argc; i++) {
  		if (argv[i][0] == '-') {
  			switch(argv[i][1]) {
  				case 'L':
  					fpl = fopen(&argv[i][2], "r");
  					break;
  				case 'R':	/* Region */
  					sscanf (&argv[i][2], "%lf/%lf/%lf/%lf", &w, &e, &s, &n);
  					west = (GMT_LONG) floor (w);	east = (GMT_LONG) ceil (e);
  					south = (GMT_LONG) floor (s);	north = (GMT_LONG) ceil (n);
  					if (west < 0) {
  						west += 360;
  						east += 360;
  					}
  					break;
  				default:
  					error = TRUE;
  					break;
  			}
  		}
  		else
  			error = TRUE;
  	}
  	if (error) {
  		fprintf(stderr,"xsetup - Prepare crossover job\n");
  		fprintf(stderr,"usage: xsetup [-L<legfile> -R<west>/<east>/<south>/<north>]\n");
  		fprintf(stderr, "	legfile is an optional list of legs to check\n");
  		fprintf(stderr, "	-R defines the region of interest. [Default is world]\n");
  		exit (EXIT_FAILURE);
  	}

   	GMT_getsharepath ("mgg", "gmt_legs", ".d", line);
 	if ((fleg = fopen (line, "r")) == NULL) {
		fprintf(stderr,"Could not open %s\n", line);
		exit (EXIT_FAILURE);
	}
   	GMT_getsharepath ("mgg", "gmt_index", ".b", line);
	if ((fbin = fopen (line, "rb")) == NULL) {
		fprintf(stderr,"Could not open %sb\n", line);
		exit (EXIT_FAILURE);
	}

	if (fpl != NULL) {
		while (fgets (line, BUFSIZ, fpl)) {
			sscanf(line, "%s", leg_used[nlegs_used]);
			nlegs_used++;
		}
		if (nlegs_used > MAXNEW) {
			fprintf(stderr, "Too many new legs, recompile!\n");
			exit (EXIT_FAILURE);
		}
		fclose(fpl);
	}

	/* Read info about each leg */

	i = 0;
	while (fgets (line, BUFSIZ,fleg)) {
		sscanf(line,"%s %" GMT_LL "d",lname, &no);
		ptr[no] = make_info (lname, no);
		legpointer[i] = no;
		ptr[no]->seq_no = i++;
		if (nlegs_used > 0)	/* use only these legs */
			ptr[no]->use = (findleg (ptr[no]->legname)) ? TRUE : FALSE;
		else	/* Use all legs */
			ptr[no]->use = TRUE;
	}
	fclose(fleg);
	n_tot_legs = i;
	n_alloc = (long)((n_tot_legs-1) / 8 + 1);
	for (i = 0; i < MAXLEGS; i++) {
		if (ptr[i]) {
			if ((ptr[i]->leglist = (unsigned char *) malloc ((size_t)n_alloc)) == NULL) {
				fprintf (stderr, "xsetup: Out of memory!\n");
				exit (EXIT_FAILURE);
			}
			for (j = 0; j < n_alloc; j++)
				ptr[i]->leglist[j] = 0;
		}
	}

	turn_on[0] = 1;
	for (i = 1; i < 8; i++)
		turn_on[i] = turn_on[i-1]*2;


	/* Start reading the index-file and write out every pair of legs that
	 * occupy the same bin
	 */

	while (fread ((void *)&bin, (size_t)4, (size_t)1,fbin) != 0) {
		skip = FALSE;
		lat = (bin / 360) - 90;
		if (lat < south || lat >= north) skip = TRUE;
		if (!skip) {
			lon = bin % 360;
			while (lon < west) lon += 360;
			if (lon >= east) skip = TRUE;
		}
		not_used = fread ((void *)&nlegs, (size_t)4, (size_t)1, fbin);
		if (skip) {
			fseek (fbin, (long int)(nlegs*4), SEEK_CUR);
			continue;
		}
		for (i = 0; i < nlegs; i++) {
			not_used = fread ((void *)&record, (size_t)4, (size_t)1, fbin);
			leg_info[i].gmt = (char )(record & 15);
			leg_info[i].leg_no = (record >> 4);
		}
		for (i = 0; i < nlegs; i++) {
			leg_1 = ptr[leg_info[i].leg_no]->seq_no;
			byte_1 = leg_1 / 8;
			bit_1 = leg_1 % 8;
			for (j = i; j < nlegs; j++) {
				leg_2 = ptr[leg_info[j].leg_no]->seq_no;
				byte_2 = leg_2 / 8;
				bit_2 = leg_2 % 8;
				ptr[leg_info[i].leg_no]->leglist[byte_2] |= turn_on[bit_2];
				ptr[leg_info[j].leg_no]->leglist[byte_1] |= turn_on[bit_1];
			}
		}
	}
	fclose(fbin);

	for (i = 0; i < MAXLEGS; i++) {
		if (ptr[i] && ptr[i]->use) {
			for (j = 0; j < n_tot_legs; j++) {
				byte_1 = j / 8;
				bit_1 = j % 8;
				ok = ptr[i]->leglist[byte_1] & turn_on[bit_1];
				if (ok && strcmp(ptr[i]->legname, ptr[legpointer[j]]->legname) <= 0)
					printf("%s %s\n", ptr[i]->legname,ptr[legpointer[j]]->legname);
			}
		}
	}

	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}

struct INFO *make_info (char *name, GMT_LONG id_no)
{
	struct INFO *new;
	if ((new = (struct INFO *) malloc(sizeof(struct INFO))) == NULL) {
		fprintf(stderr,"malloc returned NULL for make_info\n");
		exit (EXIT_FAILURE);
	}
	strcpy(new->legname,name);
	new->leg_id = id_no;
	new->use = 0;
	return (new);
}


GMT_LONG findleg (char *name)
{
	GMT_LONG left, right, mid, cmp;

	left = 0;
	right = nlegs_used-1;
	while (left <= right) {
		mid = (left + right)/2;
		cmp = strcmp(name, leg_used[mid]);
		if (cmp < 0)
			right = mid-1;
		else if (cmp > 0)
			left = mid+1;
		else
			return (1);
	}
	return (0);
}
